### PR TITLE
[feat]:[IAC-688]: Add IACM Stage Step

### PR DESCRIPTION
**Pull request by nassergonzalez**
## Describe your changes
In order to be able to execute data at the beginning of the stage and at the end of the stage.
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

Is it new task type? Is a new Stage step but I fail how this is relevant for the delegate

Did you make changes to the code in a backward-compatible manner? This is going to be used only for IACM stages which are not in prod and behind feature flags.

will the new version of the delegate receive old version task and process? I do not think that an old delegate version could run this task.

If new version of the manager receives response for the old task will it be able to handle it successfully? I do not think that an old delegate can send this task

Ng backward compatibility → did you make a change in delegate task which will be created in ng services via grpc call to manager -> no AFAIK

If the older version of ng service creates the task parameter and calls newer cg manager, will it be okay -> I do not think that this is a valid scenario

if the newer version of ng service creates the task parameter and calls older cg manager, will it be okay -> I do not think that this is a valid scenario

Did you remove an enum → this is not backward compatible yet (tracked by PL-35964: Move to name based kryo serializer for delegate/manager communication -> no AFAIK

Did you add an enum → make sure that the enum constant is added in the end of the already defined constants. If you don’t do this then you will have issues where wrong enum type will be deserialized in delegate side. -> I did

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48938)

### CODE REVIEW

**Comment by nassergonzalez** [comment]
trigger uts

File: 332-ci-manager/service/src/main/java/io/harness/ci/beans/stages/IntegrationStageStepParametersPMS.java [hunk 1]
```diff
@@ -36,12 +36,12 @@
...
 
 import java.util.List;
-import lombok.Builder;
 import lombok.Data;
+import lombok.experimental.SuperBuilder;
 import org.springframework.data.annotation.TypeAlias;
 
 @Data
-@Builder
+@SuperBuilder
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/beans/stages/IntegrationStageStepParametersPMS.java:44]: SuperBuilder is experimental

File: 310-iacm-manager/app/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java [hunk 2]
```diff
@@ -43,8 +43,9 @@ public static Map<StepType, Class<? extends Step>> getEngineSteps() {
     Map<StepType, Class<? extends Step>> engineSteps = new HashMap<>();
 
     engineSteps.put(InitializeTaskStep.STEP_TYPE, InitializeTaskStepV2.class);
-    engineSteps.put(IntegrationStageStepPMS.STEP_TYPE,
```
Comment (scottyw-harness) [inline 310-iacm-manager/app/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java:46]: Should we be removing files associated with `IntegrationStageStepPMS` or is that someone else's thing?

**Review by scottyw-harness — APPROVED** [review]

File: 332-ci-manager/service/src/main/java/io/harness/ci/beans/stages/IntegrationStageStepParametersPMS.java (hunk 1 above)
Comment (nassergonzalez) [reply 332-ci-manager/service/src/main/java/io/harness/ci/beans/stages/IntegrationStageStepParametersPMS.java:44]: I was unable to achieve this without the superbuilder annotation. If you can suggest a way to do it without the superbuilder I am happy to do it but i spent some time with Raghav and we could not find a way to do it without it

File: 310-iacm-manager/app/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java (hunk 2 above)
Comment (nassergonzalez) [reply 310-iacm-manager/app/src/main/java/io/harness/iacm/registars/IACMExecutionRegistrar.java:46]: What files?

**Comment by nassergonzalez** [comment]
trigger uts

**Comment by nassergonzalez** [comment]
trigger uts

**Review by harshjain12 — APPROVED** [review]

**Comment by nassergonzalez** [comment]
trigger codebasehash

**Comment by nassergonzalez** [comment]
trigger uts

**Pull request merged**